compiler/codegen/CodeGenRA.cpp

```diff
@@ -407,6 +407,15 @@ OMR::CodeGenerator::prepareRegistersForAssignment()
       if (registerCursor->getKind() != TR_SSR)
          registerCursor->setFutureUseCount(registerCursor->getTotalUseCount());
       km = registerCursor->getKindAsMask();
+
+      /*
+       * Sometimes real registers marked as being a TR_FPR register can be used as a TR_VRF register
+       * isFPRUsedAsVFR is used to check for this
+       * if true, RA is told is may need to assign TR_VRF as well as TR_FPR registers if a TR_FPR real register is seen
+       */
+      if (registerCursor->getKind() == TR_FPR && self()->isFPRUsedAsVFR())
+         km |= TO_KIND_MASK(TR_VRF);
+
```

Comment:
I just noticed there was a typo here. "VFR" in "isFPRUsedAsVFR" should be "VRF". I reversed the letters. I'll need to fix that.

On x, XMM physical registers can be assigned to both TR_FPR and TR_VRF virtual registers. However, each of the TR::RealRegister objects that represent each physical register can only have one register kind (register kind being something like TR_GPR, TR_FPR, TR_VRF, etc.). Right now every TR::RealRegister object that represents a XMM register is marked as being a TR_FPR register. This is true both before and after my changes since I don't make any changes to the register file.

OMR::CodeGenerator::prepareRegistersForAssignment goes through each of the TR::RealRegister objects and records the register kinds it sees. Since there are no TR::RealRegister objects marked as being a TR_VRF register, prepareRegistersForAssignment thinks that TR_VRF virtual registers do not require assignment. In reality, TR_VRF virtual registers do need a XMM register assigned to them and this fact is hidden by the fact that all of the XMM TR::RealRegister objects are marked as TR_FPR.

"isFPRUsedAsVRF" is used to indicate that the TR_FPR TR::RealRegister objects are also TR_VRF TR::RealRegister objects so TR_VRF virtual registers need to have registers assigned to them as well. "isFPRUsedAsVRF" needs to return true if TR_VRF virtual registers need to have registers assigned to them but all TR_VRF TR::RealRegister are shared with TR_FPR and are all also marked as being TR_FPR.